felix/fv/bpf_notracking_test.go

```diff
@@ -0,0 +1,125 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fv_test
+
+import (
+	"fmt"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+
+	"github.com/projectcalico/calico/felix/fv/connectivity"
+	"github.com/projectcalico/calico/felix/fv/containers"
+	"github.com/projectcalico/calico/felix/fv/infrastructure"
+	"github.com/projectcalico/calico/felix/fv/workload"
+	"github.com/projectcalico/calico/libcalico-go/lib/apiconfig"
+)
+
+var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ Felix bpf with NOTRACK feature", []apiconfig.DatastoreType{apiconfig.EtcdV3}, func(getInfra infrastructure.InfraFactory) {
+
+	if os.Getenv("FELIX_FV_ENABLE_BPF") != "true" {
+		// Non-BPF run.
+		return
+	}
+
+	const workloadTargetPort = "80"
+	var (
+		infra          infrastructure.DatastoreInfra
+		felixes        []*infrastructure.Felix
+		workloads      [2]*workload.Workload
+		cc             *connectivity.Checker
+		externalClient *containers.Container
+	)
+
+	BeforeEach(func() {
+		infra = getInfra()
+
+		opts := infrastructure.TopologyOptions{
+			FelixLogSeverity: "debug",
+			ExtraEnvVars: map[string]string{
+				"FELIX_BPFENABLED":              "true",
+				"FELIX_BPFHostConntrackBypass":  "true",
+				"FELIX_DEBUGDISABLELOGDROPPING": "true",
+			},
+		}
+		felixes, _ = infrastructure.StartNNodeTopology(2, opts, infra)
+		infra.AddDefaultAllow()
+		cc = &connectivity.Checker{}
+
+		for i := range felixes {
+			workloads[i] = workload.Run(felixes[i],
+				fmt.Sprintf("host%d-webserver", i),
+				"default",
+				fmt.Sprintf("10.65.%d.2", i),
+				workloadTargetPort, "tcp")
+			workloads[i].ConfigureInInfra(infra)
+		}
+
+		externalClient = infrastructure.RunExtClient("ext-client")
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			infra.DumpErrorData()
+		}
+
+		for _, felix := range felixes {
+			felix.Stop()
+		}
+
+		infra.Stop()
+	})
+
+	It("should allow 3rd party DNAT to workloads work", func() {
+		expectNormalConnectivity := func() {
+			cc.ResetExpectations()
+			hostIP0 := connectivity.TargetIP(felixes[0].IP)
+			cc.ExpectNone(felixes[1], hostIP0, 8080)
+			cc.ExpectNone(externalClient, hostIP0, 8080)
+			cc.ExpectNone(workloads[1], hostIP0, 8080)
+			cc.CheckConnectivity()
+		}
+
+		target := fmt.Sprintf("%s:%s", workloads[0].GetIP(), workloadTargetPort)
+
+		By("checking initial connectivity", func() {
+			expectNormalConnectivity()
+		})
+
+		By("installing 3rd party rules", func() {
+			// Install a DNAT in first felix
+			felix := felixes[0]
+
+			felix.Exec(
+				"iptables", "-t", "nat", "-A", "PREROUTING", "-p", "tcp", "-m", "tcp",
+				"--dport", "8080", "-j", "DNAT", "--to-destination", target)
+
+			cc.ResetExpectations()
+			hostIP0 := connectivity.TargetIP(felixes[0].IP)
+			cc.ExpectSome(felixes[1], hostIP0, 8080)
+			cc.ExpectSome(externalClient, hostIP0, 8080)
+			cc.ExpectSome(workloads[1], hostIP0, 8080)
```

Comment:
This is the key bit :+1:  This should not work without your fix. If you had this test as an independent commit followed by the fix commit. Running the test commit should fail and running the fix commit should pass. That is what we are testing